File: python/set3/challenge24.py
import random
import time

from basics import padding, strip_padding, xor, InvalidPaddingException
from challenge21 import MersenneTwister
import secrets


class MersenneTwisterStreamCipher:
    """
    >>> mtsc = MersenneTwisterStreamCipher(42)
    >>> pt = b'Ice Ice Baby :)'
    >>> assert pt == mtsc.decrypt(mtsc.encrypt(pt))
    """

    def __init__(self, seed):
        self.seed = seed
        self.mt = MersenneTwister(self.seed)

    def key_stream(self, length):
        return (k % 256 for k in self.mt.generate(length))

    def encrypt(self, pt: bytes):
        self.mt = MersenneTwister(self.seed)
        ppt = padding(pt, 4)
        return xor(ppt, self.key_stream(len(ppt)))

    def decrypt(self, ct: bytes):
        self.mt = MersenneTwister(self.seed)
        ppt = xor(ct, self.key_stream(len(ct)))
        try:
            return strip_padding(ppt)
        except InvalidPaddingException:
            return None


class Oracle24:

    def __init__(self):
        self.seed = secrets.randbits(16)
        self.mtsc = MersenneTwisterStreamCipher(self.seed)

    def encrypt_with_random_prefix(self, pt: bytes):
        prefix = secrets.token_bytes(secrets.randbelow(100))
        return self.mtsc.encrypt(prefix + pt)


oracle = Oracle24()


def recover_seed(ct: bytes, known: bytes):
    """
    >>> plain_text = secrets.token_bytes(16)
    >>> cipher_text = oracle.encrypt_with_random_prefix(plain_text)
    >>> recovered_seed = recover_seed(cipher_text, plain_text)
    >>> assert recovered_seed == oracle.seed
    """
    for seed in range(2**16):
        mt = MersenneTwisterStreamCipher(seed)
        cpt = mt.decrypt(ct)
        if cpt is not None and known in cpt:
            return seed


def generate_password_reset_token(seed: int):
    mt = MersenneTwisterStreamCipher(seed)
    return bytes(mt.key_stream(16))


def password_token_oracle():
    """
    Create a password token randomly in the past hour
    """
    past_time = int(time.time()) - random.randint(1, 3600)
    return generate_password_reset_token(past_time), past_time


def recover_password_token(pw_token: bytes):
    """
    >>> password_token, expected_timestamp = password_token_oracle()
    >>> recovered_timestamp = recover_password_token(password_token)
    >>> assert recovered_timestamp == expected_timestamp
    """
    current_time = int(time.time())
    for ut in range(current_time - 3700, current_time):
        if generate_password_reset_token(ut) == pw_token:
            return ut

File: kotlin/src/main/kotlin/io/github/ocirne/cryptopals/set3/Challenge24.kt
package io.github.ocirne.cryptopals.set3

import io.github.ocirne.cryptopals.Basics.InvalidPaddingException
import io.github.ocirne.cryptopals.Basics.padding
import io.github.ocirne.cryptopals.Basics.stripPadding
import io.github.ocirne.cryptopals.Basics.xor
import io.github.ocirne.cryptopals.Oracle
import io.github.ocirne.cryptopals.set3.Challenge21.MersenneTwister
import kotlin.random.Random
import kotlin.random.nextUInt

class Challenge24 {

    class MersenneTwisterStreamCipher(private val seed: UInt): Oracle {

        private fun keyStream(mt: MersenneTwister, count: Int): ByteArray {
            return mt.take(count).map { k -> (k % 256u).toByte() }.toByteArray()
        }

        fun token(): ByteArray {
            val mt = MersenneTwister(seed)
            return keyStream(mt, 16)
        }

        override fun encrypt(pt: ByteArray): ByteArray {
            val mt = MersenneTwister(seed)
            val ppt = padding(pt, 4)
            return xor(ppt, keyStream(mt, ppt.size))
        }

        override fun decrypt(ct: ByteArray): ByteArray {
            val mt = MersenneTwister(seed)
            val ppt = xor(ct, keyStream(mt, ct.size))
            return stripPadding(ppt)
        }
    }

    class RandomPrefixOracle: Oracle {

        val seed : UInt = Random.nextUInt()

        private val mtsc = MersenneTwisterStreamCipher(seed)

        override fun encrypt(pt: ByteArray): ByteArray {
            val prefix = Random.nextBytes(Random.nextInt(100))
            return mtsc.encrypt(prefix + pt)
        }
    }

    class PasswordResetTokenOracle: Oracle {

        // create a password token randomly in the last hour
        val seed: UInt = (System.currentTimeMillis() / 1000).toUInt() - Random.nextUInt(3600u)

        override fun encrypt(pt: ByteArray): ByteArray {
            throw UnsupportedOperationException()
        }

        fun generatePasswordToken(): ByteArray {
            val mtsc = MersenneTwisterStreamCipher(seed)
            return mtsc.token()
        }
    }

    class Attacker {
        fun recoverSeed(cipherText: ByteArray, known: ByteArray): UInt {
            val sKnown = String(known)
            for (seed in UInt.MAX_VALUE downTo 0u) {
                if (seed % 1000u == 0u) {
                    println(seed)
                }
                val mt = MersenneTwisterStreamCipher(seed)
                try {
                    val cpt = mt.decrypt(cipherText)
                    if (String(cpt).contains(sKnown)) {
                        return seed
                    }
                } catch (e: InvalidPaddingException) {
                    // expected for most wrong seeds
                }
            }
            throw IllegalStateException()
        }

        fun recoverTimestamp(passwordToken: ByteArray): UInt {
            val currentTime = (System.currentTimeMillis() / 1000).toUInt()
            for (ut in currentTime downTo currentTime - 3700u) {
                if (MersenneTwisterStreamCipher(ut).token().contentEquals(passwordToken)) {
                    return ut
                }
            }
            throw IllegalStateException()
        }
    }
}


